Props (#37)
Ada-Developers-Academy/core-unit-3: props/nested-components.md

# Nested Components

## Goals

> In order to increase the complexity of our React webapps, we should practice nesting components.

Reviewer: Increasing the complexity isn't the goal. But our apps _will_ become more complex, and nesting components will help us manage that.
```suggestion
In order to manage the increasing complexity of our React webapps, we should practice nesting components.
```

> Nested components share a relationship that is different from sibling components.
> 

Reviewer: I don't see anywhere that a sense of the relationship between siblings was ever established and there's no discussion of differences here, so I would cut this.
```suggestion
```

> We can apply a design pattern for components that will help us distinguish what our components do.

Reviewer: we can specifically call out separating responsibilities here

```suggestion
We can apply a design pattern for our components that will help us separate the responsibilities of the components in our applications.
```

## Format

This lesson will take the following steps:

1. Introduce two design patterns for components:
   - presentational components
   - container components

> 1. Follow Sofia's journey on creating an attendance app, applying these design patterns

Reviewer: this construction parallels the common phrasing of "journey to ..." like journey to the west, or journey to the stars, so it sounds more expected to me.
```suggestion
1. Follow Sofia's journey to create an attendance app, applying these design patterns
```

1. One instance of a `ClassInfo` component
1. One instance of a `StudentList` component

The `ClassInfo` and `StudentList` components are _siblings_ to each other.

As React webapps increase in complexity, more and more components are defined and rendered.

Deciding how to design a React webapp and its components can be challenging. We can learn about and apply two design patterns to help us:

1. Presentational components
1. Container components

These are design patterns; these labels don't affect the syntax used to define them. The difference between these two kinds of components is the nature of their responsibilities, and how we use them.

## Presentational Components

Presentational components are components that are responsible for displaying data.

Presentational components are not responsible for modifying data.

## Container Components

Container components are components responsible for:

- Using presentation components
- Holding data
- Modifying and managing data
- Sending data to presentation components

Container components often have presentation components _nested_ inside them.

## Example: `StudentList` and `Student` Components

> Sofia is a teacher creating an attendance app. Her app should display a list of students. At this moment, she starts with this implementation of a `StudentList` component in `src/components/StudentList.js`.

Reviewer: I think I've commented on "at this moment" before as being a curious mixture of immediacy in a somewhat nebulous timeframe.
```suggestion
Sofia is a teacher creating an attendance app. Her app should display a list of students. Initially, she starts with this implementation of a `StudentList` component in `src/components/StudentList.js`.
```

Author: Bahhhh yes!! I think I've been struggling with timing things (I had to edit a whole lesson after I realized I wrote that Sofia _can do things_ instead of _is doing things_ hahaha. Ughh)

Thank you again!

> (Class names have been omitted from this code snippet for clarity)

Reviewer: (previous suggestion removed)

I played around a bit more and I like the impact using blockquote for this notice has.
```suggestion
> `className` attributes have been omitted from this code snippet for clarity.
```

> (Class names have been omitted from this code snippet for clarity)
> 

Reviewer: move into the snippet as a comment so this isn't dangling out here. I tried a few stylistic things, but everything seemed out of place here.
```suggestion
```

<!-- prettier-ignore-start -->
```js
import React from 'react';
import './StudentList.css';

> 

Reviewer: moved from external remark
```suggestion

// className attributes have been omitted from this code snippet for clarity

```

const StudentList = () => {
    return (
        <section>
            <h2>Student List</h2>
            <ul>
                <li>Student A</li>
                <li>Student B</li>
                <li>Student C</li>
            </ul>
        </section>
    )
}

export default StudentList;
```
<!-- prettier-ignore-end -->

This component displays a list of students, where each student is in an `<li>` element.

> However, Sofia knows that each student list item should display a lot of data, such as their name and email address.

Reviewer: student list item parses ambiguously. is it a student-list item or a student list-item. rephrased to avoid the issue.
```suggestion
However, Sofia knows that she needs to display a lot of data about each student, such as their name and email address.
```

Sofia can apply the container and presentational component design patterns, and:

1. Create a `Student` component, which will act as a presentational component
1. Use the `Student` component in `StudentList`, which will act as a container component

### `Student` Component

> Sofia's `Student` component should display a student's name and email address. She'll create a new file, `src/components/Student.js`, and write this code:

Reviewer: switch to a present tense narration so we feel like we watch her doing these steps.
```suggestion
Sofia's `Student` component should display a student's name and email address. She creates a new file, `src/components/Student.js`, and writes this code:
```

<!-- prettier-ignore-start -->
```js
import React from 'react';

const Student = () => {
    return (

>         <li>
>             <ul>
>                 <li>Nickname: Ada</li>
>                 <li>Email: ada@dev.org</li>
>             </ul>
>         </li>

Reviewer: strongly suggest leaving the `<li>` in the list component. The student component shouldn't need to know that it's being placed in a list. If we really wanted to designate that it _must_ be part of a list, then I'd name the component something like StudentItem or StudentListItem, but even then, its render should be something like `<li><Student /></li>` with the rendering display of student being house separately.
```suggestion
        <ul>
            <li>Nickname: Ada</li>
            <li>Email: ada@dev.org</li>
        </ul>        
```

    )
}

export default Student;
```
<!-- prettier-ignore-end -->

### Rendering `Student`s in `StudentList` Component

Now, Sofia can update her `StudentList` component.

> First, she needs to import the `Student` component into `src/components/StudentList.js` with `import Student from './Student';`.
> 
> Then, she should replace the contents of the `<ul>` element.

Reviewer: same present narration style
```suggestion
First, she imports the `Student` component into `src/components/StudentList.js` with `import Student from './Student';`.

Then, she replaces the contents of the `<ul>` element.
```

> 

Reviewer: moved up from below and style with block quote which has a nice parenthetical feel to me.
```suggestion
> Imports, exports, and `className` attributes have been omitted from this code snippet for clarity.

```

<!-- prettier-ignore-start -->
```js

> const StudentList = () => {

Reviewer: moved remark from after the snippet
```suggestion
// imports, exports, and className attributes have been omitted
// from this code snippet for clarity

const StudentList = () => {
```

    return (
        <section>
            <h2>Student List</h2>
            <ul>

>                 <Student></Student>
>                 <Student></Student>
>                 <Student></Student>

Reviewer: leave the `<li>` elements here so that Student doesn't need to be aware of where it's being rendered
```suggestion
                <li><Student></Student></li>
                <li><Student></Student></li>
                <li><Student></Student></li>
```

            </ul>
        </section>
    )
}
```
<!-- prettier-ignore-end -->

> 
> (Imports, exports, and class names have been omitted from this code snippet for clarity)

Reviewer: move this remark into the snippet as a comment
```suggestion
```

> 

Reviewer: I really wanted to change `<Student></Student>` to `<Student />` but I see the former style is used a lot coming up and changing it all doesn't seem worth it. But it would be nice to add a callout about the shorthand syntax.
```suggestion

### !callout-info

## Closing Tags Are Required

React is very strict about closing tags. Even tags that are self-closing in HTML (such as `<br>` and `<hr>`) must be explicitly closed in React components. This requirement is very similar to a variant of HTML called [XHTML](https://www.w3schools.com/html/html_xhtml.asp).

<br />

Rather than forcing us to write something like `<br></br>`, we can use a shorthand: `<br />`.

<br />

We can use this same shorthand for our own components, so instead of `<Student></Student>`, we could have written `<Student />`.

### !end-callout

```

> Sofia's app now looks like this, which includes three rendered `Student` components! (This image also includes a previously implemented `ClassInfo` component.)

Reviewer: move the aside into the diagram caption
```suggestion
Sofia's app now looks like this, which includes three rendered `Student` components!
```

> ![An attendance web app with a heading "Student List," and a list of three students. All three students read "Nickname: Ada" and "Email: ada@dev.org."](../assets/props_nested-components_render.png)  
> _Fig. An attendance web app with a heading "Student List," and a list of three students. All three students read "Nickname: Ada" and "Email: ada@dev.org."_

Reviewer: revised the alt text and the caption (just minor to the alt)
alt text should describe the image itself (as this does) for folx using a screen reader to understand what the image is of
the caption should provide context for how the image relates to the surrounding text. If it's the same as the alt text, then a screen reader would end up reading the same thing twice. And for sighted folx, having the caption describe the contents of the image is strange, since they can see the image. For both cases, the caption should provide context about why the image is here.

In this case, no fancy caption is required, since it's pretty obvious that we're simply showing the updated output for reference, but we should still avoid describing the picture in words for the caption.

```suggestion
![Sophia's attendance web app with a heading "Student List," and a list of three students. All three students read "Nickname: Ada" and "Email: ada@dev.org."](../assets/props_nested-components_render.png)  
_Fig. Updated `StudentList` output. The previously implemented `ClassInfo` component is also displayed._
```

## Summary

> Sofia has applied the presentational and container component design pattern to her project.
> 
> - The `Student` component is a presentational component, responsible for display student data (name and email)

Reviewer: minor typos
```suggestion
Sofia has applied the presentational and container component design patterns to her project.

- The `Student` component is a presentational component, responsible for displaying student data (name and email)
```
